File: problems/931/931.minimum-falling-path-sum.java
/*
 * @lc app=leetcode id=931 lang=java
 *
 * [931] Minimum Falling Path Sum
 */

// @lc code=start
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        /*
         * given nxn matrix, return min sum of any falling path.
         * start from the first row and col.
         * (row, col) -> (row + 1, col - 1), (row + 1, col), (row + 1, col + 1)
         * -100 <= matrix[row][col] <= 100
         *
         * intuition:
         * for each row find the optimal minimum paths from each branch.
         * bottom-up dp
         *
         * O(N*N) time, O(N*N) space
         */

        int n = matrix.length;
        int[][] dp = new int[n][n];

        // fill base case
        for (int i = 0; i < n; i++) {
            dp[n - 1][i] = matrix[n - 1][i];
        }

        for (int row = n - 2; row >= 0; row--) {
            for (int i = 0; i < n; i++) {
                int left = i == 0 ? dp[row + 1][i] : dp[row + 1][i - 1];
                int right = i == n - 1 ? dp[row + 1][i] : dp[row + 1][i + 1];
                dp[row][i] = matrix[row][i] + Math.min(dp[row + 1][i], Math.min(left, right));
            }
        }

        int res = dp[0][0];
        for (int i = 1; i < n; i++) {
            if (dp[0][i] < res) {
                res = dp[0][i];
            }
        }

        return res;
    }
}
// @lc code=end

File: problems/931/931.minimum-falling-path-sum.py
#
# @lc app=leetcode id=931 lang=python3
#
# [931] Minimum Falling Path Sum
#


# @lc code=start
class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        """
        top-down memoization
        """

        N = len(matrix)
        dp = {}

        def dfs(i, j):
            if j < 0 or j >= len(matrix[i]):
                return float("inf")

            if i == N - 1:
                return matrix[i][j]

            if (i, j) in dp:
                return dp[(i, j)]

            res = matrix[i][j]
            res += min(dfs(i + 1, j - 1), dfs(i + 1, j), dfs(i + 1, j + 1))
            dp[(i, j)] = res
            return res

        res = float("inf")
        for j in range(N):
            res = min(res, dfs(0, j))

        return res


# @lc code=end
